segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/MetadataTransaction.java

```diff
@@ -0,0 +1,229 @@
+     * Creates a new metadata record.
+     *
+     * @param metadata metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public void create(StorageMetadata metadata) throws StorageMetadataException {
+        store.create(this, metadata);
+    }
+
+    /**
+     * Marks given record as pinned.
+     *
+     * @param metadata metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public void markPinned(StorageMetadata metadata) throws StorageMetadataException {
+        store.markPinned(this, metadata);
+    }
+
+    /**
+     * Deletes a metadata record given the key.
+     *
+     * @param key key to use to retrieve metadata.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public void delete(String key) throws StorageMetadataException {
+        store.delete(this, key);
+    }
+
+    /**
+     * Commits the transaction.
+     *
+     * @throws StorageMetadataException If transaction could not be commited.
+     */
+    public void commit() throws StorageMetadataException {
+        Preconditions.checkState(!isCommitted, "Transaction is already committed");
+        Preconditions.checkState(!isAborted, "Transaction is already aborted");
+        store.commit(this);
+        isCommitted = true;
+    }
+
+    /**
+     * Commits the transaction.
+     *
+     * @param lazyWrite true if data can be written lazily.
+     * @throws StorageMetadataException If transaction could not be commited.
+     */
+    public void commit(boolean lazyWrite) throws StorageMetadataException {
+        Preconditions.checkState(!isCommitted, "Transaction is already committed");
+        Preconditions.checkState(!isAborted, "Transaction is already aborted");
+        store.commit(this, lazyWrite);
+        isCommitted = true;
+    }
+
+    /**
+     * Commits the transaction.
+     *
+     * @param lazyWrite      true if data can be written lazily.
+     * @param skipStoreCheck true if data is not to be reloaded from store.
+     * @throws StorageMetadataException If transaction could not be commited.
+     */
+    public void commit(boolean lazyWrite, boolean skipStoreCheck) throws StorageMetadataException {
+        Preconditions.checkState(!isCommitted, "Transaction is already committed");
+        Preconditions.checkState(!isAborted, "Transaction is already aborted");
+        store.commit(this, lazyWrite, skipStoreCheck);
+        isCommitted = true;
+    }
+
+    /**
+     * Aborts the transaction.
+     *
+     * @throws StorageMetadataException If transaction could not be commited.
+     */
+    public void abort() throws StorageMetadataException {
+        Preconditions.checkState(!isCommitted, "Transaction is already committed");
+        Preconditions.checkState(!isAborted, "Transaction is already aborted");
+        isAborted = true;
+        store.abort(this);
```

Comment:
If there is an error in `store.abort`, is the transaction to be considered *aborted* ?

Reply:
Data is written to underlying store only when transaction is committed. Until then the underlying store or cache has no knowledge of changes in your transaction. So if you fail to commit, it is an implicit abort. In fact the `abort` implementation in `BaseMetadataStore` is empty. There is no action required. `TableBasedMetadataStore` or `InMemoryMetadataStore` do not override it. If some derived class overrides it then it may throw an error. But transaction is already marked `isAborted = true;`

Reply:
So transactions are only in memory until you commit them. and if you reboot the service you will lose your transaction data. is that correct ?

Reply:
Yes. That is correct.
These are short lived transactions. ChunkManager should either succeed in committing or in case of failed/rebooted service leave no side effect. Therefore the next instance of the service will not have to deal with any partial changes to metadata.